src/ray/raylet/node_manager.cc

```diff
@@ -1454,7 +1470,7 @@ void NodeManager::HandleObjectLocal(const ObjectID &object_id) {
     // (See design_docs/task_states.rst for the state transition diagram.)
     local_queues_.MoveTasks(ready_task_id_set, TaskState::WAITING, TaskState::READY);
     // New ready tasks appeared in the queue, try to dispatch them.
-    DispatchTasks();
+    DispatchTasks(nullptr);
```

Comment:
We could actually call `DispatchTasks` directly on the tasks in the `ready_task_id_set` instead of calling it on the entire ready queue.

Reply:
Ok, let me take a look

Reply:
Hmm it shouldn't be too hard to do, we just can't use the `local_queues_.MoveTasks(...)` method for now... but you can do something like this:

```c++
// First filter out the tasks that should not be moved to READY.
local_queues_.FilterState(ready_task_id_set, TaskState::BLOCKED);
local_queues_.FilterState(ready_task_id_set, TaskState::DRIVER);
// Optionally, make sure that the remaining tasks are WAITING.
const auto ready_task_id_set_copy = ready_task_id_set;
local_queues_.FilterState(ready_task_id_set_copy, TaskState::WAITING);
// Queue and dispatch the tasks that are ready to run.
auto ready_tasks = local_queues_.RemoveTasks(ready_task_id_set)
local_queues_.QueueReadyTasks(ready_tasks);
DispatchTasks(ready_tasks);
```

This isn't the cleanest, but I think we need to be careful here because this could definitely lead to O(N^2) work to schedule tasks again (e.g., if many tasks have dependencies). A good performance test would be something like, submit 10k tasks with no dependencies that exit immediately, then submit 10k more tasks that sleep for a while, each dependent on one task from the previous round.

Reply:
anything can be hacked, the question is whether it's desirable. We don't expose task getters in the SchedulingQueue API by design. The hope was to keep all the tasks in the queues. I wonder if it would be cleaner to have `DispatchTasks` take a set of task ids instead. Then it can `RemoveTasks` on them internally. If the set is empty, the assumption is all ready tasks. Is there a good reason to pass a task references instead?

Reply:
I see... What is `ready_task_id_set_copy` used for? Did you want to check `RAY_CHECK(ready_task_id_set_copy.empty())`?

Reply:
Ok, did it and resubmit it. Note that DispatchTasks() takes a list of tasks not a vector of tasks so there is a little bit of added ugliness. 

Also, note that the implicit assumption here is that the number of tasks whose dependencies are fulfilled is not that large as we do several traversals of this list to filter tasks out.

Reply:
@istoica and I talked offline and we think it makes sense to have `DispatchTasks` still take in a `const &` to the list of tasks that may be dispatched. To avoid the copy of the task list (which is already an issue, #3195), `DispatchTasks` should return the list of task IDs that were dispatched and can be removed from the `READY` queue instead of removing the tasks directly in that method.

Reply:
Done, and pushed the changes.